File: isma-intg-api/src/main/java/ru/nstu/isma/intg/api/calcmodel/AlgebraicEquationCalculator.java
package ru.nstu.isma.intg.api.calcmodel;

/**
 * @author Maria Nasyrova
 * @since 04.10.2015
 */
public class AlgebraicEquationCalculator implements AlgebraicEquationResultProvider {

    private final double[] y;
    private final Double[] calcResults;
    private final AlgebraicEquation[] algebraicEquations;

    public AlgebraicEquationCalculator(double[] y, AlgebraicEquation[] algebraicEquations) {
        this.y = y;
        this.calcResults = new Double[algebraicEquations.length];
        this.algebraicEquations = algebraicEquations;
    }

    @Override
    public double getValue(int index) {
        if (calcResults[index] == null) {
            calcResults[index] = algebraicEquations[index].apply(y, this);
        }
        return calcResults[index];
    }

    public double[] getValues() {
        double[] values = new double[calcResults.length];
        for (int i = 0; i < values.length; i++) {
            values[i] = getValue(i);
        }
        return values;
    }
}

File: isma-solver/api/src/main/kotlin/ru/nstu/isma/intg/api/calcmodel/AlgebraicEquationCalculator.kt
package ru.nstu.isma.intg.api.calcmodel

import ru.nstu.isma.compiler.hsm.jvm.calcmodel.AlgebraicEquation

class AlgebraicEquationCalculator(
    private val algebraicEquations: Array<AlgebraicEquation>
) {
    fun apply(y: DoubleArray): DoubleArray {
        val rhsForDe = DoubleArray(algebraicEquations.size)
        for (i in algebraicEquations.indices) {
            rhsForDe[i] = algebraicEquations[i].apply(y)
        }
        return rhsForDe
    }
}

